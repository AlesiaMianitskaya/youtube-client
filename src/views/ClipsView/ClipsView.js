export default class ClipsView {
  constructor() {
    const content = document.createElement('div');
    content.className = 'clips';
    document.body.appendChild(content);
    content.addEventListener('scroll', this.onScroll);
    this.content = content;
  }

  onMore = undefined;

  onScroll = e => {
    const el = e.target;
    if (el.scrollWidth === el.clientWidth + el.scrollLeft) {
      if (this.onMore) {
        this.onMore();
      }
    }
  };

  clear = () => {
    this.content.innerHTML = '';
  };

  renderEntry = data => {
    const entry = document.createElement('div');
    entry.setAttribute('class', 'clip-item');
    entry.setAttribute('style', `background-image: url(${data.preview})`);
    entry.innerHTML = `
      <div>
        <a target='_blank' href="'https://www.youtube.com/watch?v=${data.id}">
          ${data.title}
        </a>
        <h5 class='author-name'><i class='material-icons'>face</i>
        ${data.author}
        </h5>
        <p class='date'><i class="material-icons">today</i>
        ${data.publishedAt.slice(0, -14)}
        </p>
        <p class='views'><i class='material-icons'>visibility</i>
        ${data.views}
        </p>
        <p class='description'>${data.description}</p>
      </div>   
    `;
    this.content.appendChild(entry);
    ClipsView.srcollClips();
  };

  static srcollClips() {
    const slider = document.querySelector('.clips');
    let isDown = false;
    let startX;
    let scrollLeft;

    slider.addEventListener('mousedown', e => {
      isDown = true;
      slider.classList.add('active');
      startX = e.pageX - slider.offsetLeft;
      scrollLeft = slider.scrollLeft;
    });
    slider.addEventListener('mouseleave', () => {
      isDown = false;
      slider.classList.remove('active');
    });
    slider.addEventListener('mouseup', () => {
      isDown = false;
      slider.classList.remove('active');
    });
    slider.addEventListener('mousemove', e => {
      if (!isDown) return;
      e.preventDefault();
      const x = e.pageX - slider.offsetLeft;
      const walk = (x - startX) * 2;
      slider.scrollLeft = scrollLeft - walk;
    });
  }
}
