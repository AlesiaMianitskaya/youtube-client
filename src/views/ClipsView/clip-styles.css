@import url('https://fonts.googleapis.com/icon?family=Material+Icons');
.clips {
  height: 420px;
  width: 1220px;
  margin: 70px auto;
  overflow-x: scroll;
  overflow-y: hidden;
  white-space: nowrap;
}

.clip-item {
  display: inline-block;
  width: 250px;
  height: 400px;
  margin: 0px 30px;
  vertical-align: middle;
  background-color: #f2d2ea;
  text-align: center;
  white-space: normal;
  overflow-wrap: break-word;
  border: 1px solid #3c3d3a;
  border-radius: 5px;
  font-size: 1rem;
  overflow: hidden;

  background-position: center 0px;
  background-size: 100% 45%;
  background-repeat: no-repeat;
}

.clip-item div {
  height: 225px;
  margin: 160px 0px 0px 10px;
  text-align: left;
}

.clips img {
  height: 40%;
  width: 90%;
}

.clips a {
  background: white;
  display: block;
  margin: 50px 10px 0px 0px;
  color: #060606;
  font-weight: 500;
  text-align: center;
}

.author-name {
  margin: 10px 0px;
  font-size: 1rem;
}

.description {
  font-size: 0.9rem;
  font-style: italic;
  font-weight: 300;
  line-height: 1.4;
}

.clips.active {
  cursor: grabbing;
}

.material-icons {
  margin-right: 15px;
  vertical-align: middle;
}

/* 
  ##Device = Tablets, Ipads (portrait)
  ##Screen = from 768px to 1024px
*/
@media (min-width: 768px) and (max-width: 1024px) {
  .clips {
    width: 630px;
    margin: 50px auto;
  }
}

/* 
  ##Device = Low Resolution Tablets, Mobiles
  ##Screen = from 481px to 767px
*/
@media (min-width: 481px) and (max-width: 767px) {
  .clips {
    width: 310px;
    margin: 30px auto;
  }
}

/* 
  ##Device = Most of the Smartphones Mobiles
  ##Screen = from 320px to 479px
*/
@media (min-width: 320px) and (max-width: 480px) {
  .clips {
    width: 300px;
    margin: 10px auto;
  }
}
