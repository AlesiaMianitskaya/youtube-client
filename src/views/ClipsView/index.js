import ClipsView from './ClipsView';

import './clip-styles.css';

export default ClipsView;
