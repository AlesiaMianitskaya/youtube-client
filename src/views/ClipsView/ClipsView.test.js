import ClipsView from './ClipsView';

describe('ClipsView.renderEntry', () => {
  it('Should be render clips correctly', () => {
    const clipView = new ClipsView();
    const clip = [
      {
        title: 'JavaScript',
        description: 'Web',
        author: 'Web Developer',
        publishedAt: '2018-09-06T10:47:49.000Z',
        preview: 'https:/img.jpg',
        views: '3000',
      },
    ];
    clip.forEach(clipView.renderEntry);
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
