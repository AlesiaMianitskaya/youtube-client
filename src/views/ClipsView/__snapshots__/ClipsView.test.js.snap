// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClipsView.renderEntry Should be render clips correctly 1`] = `
"<div class=\\"clips\\"><div class=\\"clip-item\\" style=\\"background-image: url(https:/img.jpg)\\">
      <div>
        <a target=\\"_blank\\" href=\\"'https://www.youtube.com/watch?v=undefined\\">
          JavaScript
        </a>
        <h5 class=\\"author-name\\"><i class=\\"material-icons\\">face</i>
        Web Developer
        </h5>
        <p class=\\"date\\"><i class=\\"material-icons\\">today</i>
        2018-09-06
        </p>
        <p class=\\"views\\"><i class=\\"material-icons\\">visibility</i>
        3000
        </p>
        <p class=\\"description\\">Web</p>
      </div>   
    </div></div>"
`;

exports[`SearchView.prototype.renderEntry Should be render clips correctly 1`] = `
"<div class=\\"clips\\"><div class=\\"clip-item\\" style=\\"background-image: url(https:/img.jpg)\\">
      <div>
        <a target=\\"_blank\\" href=\\"'https://www.youtube.com/watch?v=undefined\\">
          JavaScript
        </a>
        <h5 class=\\"author-name\\"><i class=\\"material-icons\\">face</i>
        Web Developer
        </h5>
        <p class=\\"date\\"><i class=\\"material-icons\\">today</i>
        2018-09-06
        </p>
        <p class=\\"views\\"><i class=\\"material-icons\\">visibility</i>
        3000
        </p>
        <p class=\\"description\\">Web</p>
      </div>   
    </div></div>"
`;
